Add snapshot regression tests using Playwright

rgieseke opened the pull request:
To have some end to end testing as discussed in #244 it might be an option to use Playwright's snapshot tests of the rendered components. This would ensure that the charts display.

Playwright docs: https://playwright.dev/docs/test-snapshots

With Playwright there are of course many more ways to add tests.

https://svelte.dev/docs/svelte/testing#E2E-tests-using-Playwright

There is also (experimental) component testing support: 

https://playwright.dev/docs/test-components

As for the tests, I'm not yet sure how stable this is between platforms (e.g. Mac vs. Linux) and different Chromium versions, so I'm not sure whether it's helpful to add it to CI. In any case it's quite useful to run after making changes. From some brief testing it seems to me that a Svelte Migrate run doesn't cause any problems (it does not yet test interactivity though).

The visual diffs are run against a local preview server and the current command invokes a `npm run build` unless a preview server is already running on 4173.

The command to run the tests is 

```
npx playwright test
```  

To run an individual test one can run

```
npx playwright test -g "Snapshot for /example/CirclePackForce"
```

To check how errors would look, I did some changes, e.g. change Bar color or Axis `dy` setting.

![image](https://github.com/user-attachments/assets/04d8b498-e80b-4b8b-b93f-4aa806784620)


![image](https://github.com/user-attachments/assets/9f097a4c-55bd-483e-8270-83efaa237867)

mhkeller commented:
That's very neat!!

mhkeller commented:
Very cool. I'm still reading about how this all works together. I merged #256 so that should help move this along. Thanks for your patience! It would be helpful to add the test generation and playwright test running commands to the package.json scripts. Maybe keep them separate as `test:playwright", change the current mocha command to `test:lib` and then `test: npm run test test:lib && npm run test:playwright`. I'd be curious to see how they work in CI. It would be nice if it were part of the automatic checks.

rgieseke commented:
Yes, this should definitely run in CI, the text-based ARIA tests should be no problem. My concern was that the screenshot generation might not be stable between local generation of snapshots and whatever is run in CI (and will potentially change). By only using the chart itself I tried to minimize that a bit. Maybe in the Axis*.svelte charts the inputs should also be removed from the screenshot as their design varies between browsers.

The force layouts are also potentially problematic and might require longer timeouts in CI or between platforms. Have you experimented with speeding them up from Layer Cake? Maybe the [alpha](https://d3js.org/d3-force/simulation#simulation_alpha) parameter could be exposed as a prop and set to run faster during tests.  

The other issue is the need of local preview server and a fresh build (or re-using an already running preview server). Maybe i'll add two commands `build-and-test:playwright` and only `test:playwright`. 
It depends on whether one is working on the tests or on the code but having to wait for a full build probably shouldn't be the default.

rgieseke commented:
When running Playwright in CI there are a few minor pixel differences. This could be worked around by setting a higher max pixel difference threshold.

You can download the created artifact with the report from:
https://github.com/mhkeller/layercake/actions/runs/14081936587/artifacts/2822882323

mhkeller commented:
Made a couple of tweaks but this is looking very cool. 

> To update the YAML snapshots I think it's necessary to delete them and run
> 
> ```
> npx playwright test --update-snapshots
> ```

I added this as an npm script. Does that look right to you?

```
"test:init:snapshots": "rm -rf test/e2e/component-properties-table.spec.js-snapshots && playwright test --update-snapshots",
```

 What's the command for creating the screenshots?

It seems like you fixed the force-directed issue by adding a greater timeout there. For the artifacts, it seems like it only uploads images when the test fails, is that right?

The only one that's failing is the Calendar component, which looks quite off 
<img width="1075" alt="Screenshot 2025-04-08 at 21 44 26" src="https://github.com/user-attachments/assets/333437cf-fac5-473b-86ff-1cfc1085be82" />

I think adding a max pixel different would be good in addition to the ratio, currently set at 2.5 percent. Will need to test on what a good amount is there.

rgieseke commented:
> > To update the YAML snapshots I think it's necessary to delete them and run
> > ```
> > npx playwright test --update-snapshots
> > ```
> 
> I added this as an npm script. Does that look right to you?
> 
> ```
> "test:init:snapshots": "rm -rf test/e2e/component-properties-table.spec.js-snapshots && playwright test --update-snapshots",
> ```

I think I was mistaken, just running `playwright test --update-snapshots` should be enough.
Guess I didn't check for a changed file modification date.

> 
> What's the command for creating the screenshots?
> 
That should run both.

> It seems like you fixed the force-directed issue by adding a greater timeout there. For the artifacts, it seems like it only uploads images when the test fails, is that right?

Hm, I think this should only not do the artifact upload when the run is actually ['cancelled'](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#cancelled).  

https://github.com/mhkeller/layercake/blob/b1f287076200dc11479ab6f66356a8da084b7f24/.github/workflows/playwright.yml#L21

> The only one that's failing is the Calendar component, which looks quite off 

See #265 - I hope have fixed this now. Once this is correct, hopefully there will be no big differences between local and CI (likely using UTC).

rgieseke commented:
I've updated the screenshots to use the default filename from playwright which includes the platform, e.g.

`test/e2e/chart-screenshots.spec.js-snapshots/Snapshot-for-components-Annotations-html-svelte-1-chromium-linux.png`

I think this makes it clearer to understand potential differences when tests are run on a different platform.

rgieseke commented:
> > I added this as an npm script. Does that look right to you?
> > ```
> > "test:init:snapshots": "rm -rf test/e2e/component-properties-table.spec.js-snapshots && playwright test --update-snapshots",
> > ```
> 
> I think I was mistaken, just running `playwright test --update-snapshots` should be enough. Guess I didn't check for a changed file modification date.
> 

Correction: only non matching ones will be updated

> Running tests with the --update-snapshots flag will update snapshots that did not match. Matching snapshots will not be updated.

[Docs](https://playwright.dev/docs/aria-snapshots#updating-snapshots-with-playwrighttest-and-the---update-snapshots-flag)

I've simplified the `test:init:snapshot` command in `package.json`

mhkeller commented:
Just merged https://github.com/mhkeller/layercake/pull/265, so I'm guessing we have to re-run the snapshots and screenshots and then that should resolve?

rgieseke commented:
> Just merged #265, so I'm guessing we have to re-run the snapshots and screenshots and then that should resolve?

I did merge `main` and did `npm run test:init:snapshots`. The calendar tests are now passing, the failing tests are due to minor differences in `AreaStacked` rendering, see below (copied from the report artifact):

> Error: expect(locator).toHaveScreenshot(expected) 321 pixels (ratio 0.01 of all image pixels) are different.
![image](https://github.com/user-attachments/assets/30899c7f-f301-400d-9b3b-3e62b36b8b19)

rgieseke commented:
I made the following changes to get the tests to pass:

- pixel ratio is only set in `playwright.config.js` (and overwritten for force layout tests) 
- i removed the fixed pixel count `maxDiffPixels` (I think these were causing e.g. AreaStacked with 351 pixels off to fail)

mhkeller commented:
Should this one also be a draft or is this ready for a review?

mhkeller commented on test/e2e/component-properties-table.spec.js-snapshots/Property-table-content-for-components-Arrows-svelte-1.aria.yml:
```diff
@@ -0,0 +1,33 @@
+        - code: "'.chart-container'"
+      - cell "no"
+      - cell "The class name / CSS selector of the parent element of the <LayerCake> component. This is used to crawl the DOM for the text annotations.":
+        - code: <LayerCake>
```
This is auto-generated from playwright ya? So I feel like it's not fixable

rgieseke replied:
Yes, but I think it shouldn't matter (if you run `npm run format` the tests still pass even though the indentation changed, playwright does the comparison on a parsed snapshot).

mhkeller commented:
For fun, I just ran this new Copilot review. It may be garbage. Kinda curious.

rgieseke commented:
> Should this one also be a draft or is this ready for a review?

I thought it was ready but I just realized that the current pixel difference ratio is not yet right. If one for example changes the font size in `Sankey.svelte` a bit this is not detected. Maybe this has to do with the amount of pixels in total in the screenshots.
A small line width change for example in `Radar.svelte` is showing a diff.

Otherwise the test suite has been quite useful to me. All the small JSDoc issues I found through the tests when experimenting with Svelte 5 conversions and then running the tests.

> For fun, I just ran this new Copilot review. It may be garbage. Kinda curious.

Interesting! The Yaml is auto-generated but `npm run format` will also change the yaml files formatting. I think it shouldn't matter.

I merged `main` with the JSDoc fixes and as expected the changes in `Radar.svelte` and `Column.svelte` are detected. For `Arrows.svelte` there is no change as the current code only parses the JSDoc-strings.

I'll run `npm run test:init:snapshots` to get the new snapshots for `Radar` and `Column` (maybe it should be called `test:update:snapshots`).

rgieseke commented:
I think the problem with the screenshot comparison might be that white/transparent background pixels don't 'count' as image pixels? An image with more visible pixels like `Sankey` compared with `Radar` thus can have more different pixels before the ratio triggers an error.
The library used is [pixelmatch](https://github.com/mapbox/pixelmatch)

rgieseke commented:
I removed the visual screenshot comparisons and will start a separate PR for these. 

Then this PR is ready for review (and #270 could be reviewed afterwards).

mhkeller commented on test/e2e/component-properties-table.spec.js:
```diff
@@ -0,0 +1,24 @@
+				'BarStacked.svelte.html',
+				'ColumnStacked.svelte.html',
+				'GroupLabels.html.svelte.html'
+			].includes(i)
```
My only thought is I wonder if we can auto-detect this but this isn't necessary to solve now.

rgieseke replied:
That's a good idea to future-proof that. I'll take a look. Should be possible somehow to skip the test if there is no match for `#params-table`.

mhkeller commented:
Thanks again for all of your work on this. It's a huge lift and I appreciate the time you've put into it!
